qiita_db/analysis.py

```diff
@@ -958,7 +983,78 @@ def _build_biom_tables(self, grouped_samples, rename_dup_samples=False):
                     new_table.to_hdf5(
                         f, "Generated by Qiita, analysis id: %d, info: %s" % (
                             self._id, label))
-                biom_files.append((data_type, biom_fp))
+
+                # post_processing_cmds are a list of commands to run on the
+                # final BIOM. The order of operations is list-order. Each
+                # element of the list is a dictionary containing the Conda env
+                # to use, the script to run, and a dictionary of parameters.
+                if post_processing_cmds:
+                    # assuming all commands require fragments, obtain
+                    # fragments once, instead of for every cmd.
+                    features = load_table(biom_fp).ids(axis='observation')
+                    features = list(features)
+                    fragments = Archive.retrieve_feature_values(
+                        archive_merging_scheme=archive_merging_scheme,
+                        features=features)
+
+                    # remove fragments that SEPP could not match
+                    fragments = {f: loads(fragments[f])
+                                 for f, plc
+                                 in fragments.items()
+                                 if plc != ''}
+
+                    # since biom_fp uses base_fp as its location, assume it's
+                    # suitable for other files as well.
+                    output_dir = base_fp
+
+                    fp_fragments = '%s/fragments.json' % output_dir
+
+                    with open(fp_fragments, 'w') as out_file:
+                        dump(fragments, out_file)
+
+                    for cmd in post_processing_cmds:
+                        # assume fragments file is passed as:
+                        # --fp_fragments=<path_to_fragments_file>
+                        # assume output dir is passed as:
+                        # --output_dir=<path_to_output_dir>
+
+                        # concatenate any other parameters into a string
+                        params = ' '.join(["%s=%s" % (k, v) for k, v in
+                                          cmd['script_params'].items()])
+
+                        # append fragments file and output dir parameters
+                        params = "%s --fp_fragments=%s --output_dir=%s" %\
```

Comment:
Can you change the name of the parameter from `--fp_fragments=` to `--fp_archive` so it's more generic; at the end of the day this can be fragments or anything else the plugin stores in the database/archive

Reply:
No problem - changes already made to qiita and qp_deblur. One thing though, the command line util that strips fragments from the biom that aren't in the tree takes --fp_phylogeny and --fp_biom as options; do you think they might be too specific as well?

Originally, I thought a chain of commands would each in turn iterate over the same biom file, but since the second command will need a tree and a biom file as input, the database entry will need to be updated with the fp_s to the data, or the second command will need to assume the name and location of the inputs.

Reply:
I think those names internally in deblur are fine but you are right about those here. Also all the comments that mention phylogeny vs. extra file for analysis ...